### ISSUE TITLE
Union: Unpack is not allowed in this context

### ISSUE THREAD

**Issue by nernst** [microsoft/pyright#7416]
## Environment data

-   Language Server version: v2024.2.105 (pre-release)
-   OS and version: Ubuntu 20.04.6 LTS
-   Python version (& distribution if applicable, e.g. Anaconda): CPython 3.12.2

## Code Snippet

```python
from typing import Union
types = (int, str)
Union[*types]  # Unpack is not allowed in this context
```

## Repro Steps

1. XXX

## Expected behavior

No errors should be reported.

## Actual behavior

Pylance reports "Unpack is not allowed in this context"

## Logs

```
XXX
```

**Comment by erictraut** [comment]
@nerst, can you explain what you're trying to do here? Are you using `Union` in a value expression? I ask because the unpack operator `*` is not allowed in a type expression (i.e. a type annotation).

**Comment by erictraut** [comment]
I'm going to close this because I think pyright is working correctly here. If you would like to provide more information about what you're trying to do, I'm happy to suggest an alternative solution that doesn't violate the typing spec.

**Comment by ntjess** [comment]
@erictraut Here is a contrived example based on my use case. I have a registry users can inform of their own classes, which is used to deserialize json data through Pydantic.

I would like Pydantic's `TypeAdapter` to operate on a Union of the dynamically provisioned class types, like so:

<details>

```python
import typing as t
from pydantic import BaseModel, TypeAdapter

BaseModelType_T = t.TypeVar("BaseModelType_T", bound=t.Type[BaseModel])
REGISTRY = []

def register(cls: BaseModelType_T) -> BaseModelType_T:
    REGISTRY.append(cls)
    return cls

class LoadedData(t.TypedDict):
    image_path: str
    mime_type: str

class SavedData(t.TypedDict):
    success: bool

@register
class Event(BaseModel):
    type: t.Literal["Event"] = "Event"
    data: t.Any

@register
class LoadEvent(BaseModel):
    type: t.Literal["LoadEvent"] = "LoadEvent"
    data: LoadedData

@register
class SaveEvent(BaseModel):
    type: t.Literal["SaveEvent"] = "SaveEvent"
    data: SavedData

data = [
    {"type": "Event", "data": 42},
    {"type": "LoadEvent", "data": {"image_path": "foo.png", "mime_type": "image/png"}},
    {"type": "SaveEvent", "data": {"success": True}},
]

ta = TypeAdapter(t.Sequence[t.Union[*REGISTRY]])
ta.validate_python(data)
```

</details>

This program correctly outputs the deserialized classes, but Pyright complains of the type unpacking. Is there a different recommended approach to achieve the same result?

**Comment by erictraut** [comment]
Thanks for the code sample, @ntjess. I can see that you're using `Union` outside of a type expression context. In this case, it should be treated like any other value expression, and type expression rules should not apply. That means pyright shouldn't generate an error in this situation. This will be addressed in the next release.

**Comment by erictraut** [comment]
This is addressed in pyright 1.1.384
